# Analiza i obróbka danych - branded foods

#### tj czyli te "markowe" jak np chipsy lays, żelki haribo itp

## 1. Dane
Dane zostały pobrane z strony USDA: https://fdc.nal.usda.gov/download-datasets
Rozpakowany katalog z plikami CSV: `USDA_branded_foods` z branded foods ma 2.9 GB. W obróbce danych posłyżymy się dokumentacją z USDA na temat tego, co jaki plik zawiera oraz jak interpretować te dane.  
Tym razem załadujemy do DF'a pliki CSV zamiast json, ale 
i tak finalnie obrobione produkty będą w formcacie json dokładnie tak jak te z "foundational foods", oto 2 przykładowe rekordy dla przypomnienia:
```json
  {
    "usdaId":321358,
    "name":"Hummus, commercial",
    "unit":{
      "usdaUnitId":2048,
      "name":"gram",
      "symbol":"g"
    },
    "gramsPerUnit":1.0,
    "kcal100g":229.0,
    "protein100g":7.35,
    "carbs100g":14.9,
    "fat100g":17.1
  },
  {
    "usdaId":321360,
    "name":"Tomatoes, grape, raw",
    "unit":{
      "usdaUnitId":1000,
      "name":"cup",
      "symbol":"cup"
    },
    "gramsPerUnit":152.0,
    "kcal100g":27.0,
    "protein100g":0.83,
    "carbs100g":5.51,
    "fat100g":0.63
  }
```

#### Analiza plików .csv

W pobranym katalogu z USDA jest 9 różnych plików `.csv`

 - branded_food.csv
 - food.csv
 - food_attribute.csv
 - food_attribute_type.csv
 - food_nutrient.csv
 - food_update_log_entry.csv
 - measure_unit.csv
 - microbe.csv
 - nutrient.csv

Wypiszemy zawartość 3 pierwszych rekordów z każdego z nich, aby zobaczyć z czym w ogóle pracujemy

In [4]:
import pandas as pd
from pathlib import Path

DATA_DIR = Path(r"C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\USDA_branded_foods")

csv_files = sorted([p for p in DATA_DIR.iterdir() if p.suffix.lower() == ".csv"])

print("Znalezione CSV:")
for f in csv_files:
    print(" -", f.name)

print("\n--- Podgląd pierwszych 3 wierszy każdego pliku ---\n")

for file in csv_files:
    print(f"\n### {file.name} ###")
    try:
        df = pd.read_csv(file, nrows=3, low_memory=False)
        display(df)
    except Exception as e:
        print("Błąd przy wczytywaniu:", e)


Znalezione CSV:
 - branded_food.csv
 - food.csv
 - food_attribute.csv
 - food_attribute_type.csv
 - food_nutrient.csv
 - food_update_log_entry.csv
 - measure_unit.csv
 - microbe.csv
 - nutrient.csv

--- Podgląd pierwszych 3 wierszy każdego pliku ---


### branded_food.csv ###


,fdc_id,brand_owner,brand_name,subbrand_name,gtin_upc,ingredients,not_a_significant_source_of,serving_size,serving_size_unit,household_serving_fulltext,...,data_source,package_weight,modified_date,available_date,market_country,discontinued_date,preparation_state_code,trade_channel,short_description,material_code
0,1105904,Richardson Oilseed Products (US) Limited,NaN,NaN,27000612323,Vegetable Oil,NaN,15.0,ml,1 Tbsp (15 ml),...,GDSN,NaN,2020-10-02,2020-11-13,United States,NaN,NaN,NaN,NaN,NaN
1,1105905,CAMPBELL SOUP COMPANY,NaN,NaN,51000198808,"INGREDIENTS: BEEF STOCK, CONTAINS LESS THAN 2%...",NaN,240.0,ml,Amount per serving,...,GDSN,NaN,2020-09-12,2020-11-13,United States,NaN,NaN,NaN,NaN,NaN
2,1105906,CAMPBELL SOUP COMPANY,NaN,NaN,51000213273,"INGREDIENTS: CLAM STOCK, POTATOES, CLAMS, CREA...",NaN,440.0,g,PER CONTAINER,...,GDSN,NaN,2020-09-01,2020-11-13,United States,NaN,NaN,NaN,NaN,NaN



### food.csv ###


,fdc_id,data_type,description,food_category_id,publication_date,market_country,trade_channel,microbe_data
0,1105904,branded_food,WESSON Vegetable Oil 1 GAL,NaN,2020-11-13,United States,NaN,"[{""min_value"":null,""uom"":null,""microbe_method""..."
1,1105905,branded_food,SWANSON BROTH BEEF,NaN,2020-11-13,United States,NaN,"[{""min_value"":null,""uom"":null,""microbe_method""..."
2,1105906,branded_food,CAMPBELL'S SLOW KETTLE SOUP CLAM CHOWDER,NaN,2020-11-13,United States,NaN,"[{""min_value"":null,""uom"":null,""microbe_method""..."



### food_attribute.csv ###


,id,fdc_id,seq_num,food_attribute_type_id,name,value
0,1009437,1105904,NaN,NaN,Ingredients,3
1,1009439,1105905,NaN,NaN,Ingredients,3
2,1009438,1105905,NaN,NaN,Nutrient Removed,6



### food_attribute_type.csv ###


,id,name,description
0,998,Update Log,Changes that were made to this food
1,999,Attribute,Generic attributes
2,1000,Common Name,Common names associated with a food.



### food_nutrient.csv ###


,id,fdc_id,nutrient_id,amount,data_points,derivation_id,min,max,median,footnote,min_year_acquired
0,13706927,1105904,1257,0.00,NaN,71,NaN,NaN,NaN,NaN,NaN
1,13706930,1105904,1293,53.33,NaN,71,NaN,NaN,NaN,NaN,NaN
2,13706926,1105904,1253,0.00,NaN,75,NaN,NaN,NaN,NaN,NaN



### food_update_log_entry.csv ###


,id,description,last_updated
0,1105904,WESSON Vegetable Oil 1 GAL,2020-11-13
1,1105905,SWANSON BROTH BEEF,2020-11-13
2,1105906,CAMPBELL'S SLOW KETTLE SOUP CLAM CHOWDER,2020-11-13



### measure_unit.csv ###


,id,name
0,1000,cup
1,1001,tablespoon
2,1002,teaspoon



### microbe.csv ###


,id,foodId,method,microbe_code,min_value,max_value,uom
0,1,2331974,AEROBIC_PLATE_COUNT,LACTICASEIBACILLUS_PARACASEI,10000000,NaN,NaN
1,2,2332523,AEROBIC_PLATE_COUNT,LACTICASEIBACILLUS_PARACASEI,10000000,NaN,NaN
2,3,2336972,AEROBIC_PLATE_COUNT,LACTICASEIBACILLUS_PARACASEI,10000000,NaN,X_CFG



### nutrient.csv ###


,id,name,unit_name,nutrient_nbr,rank
0,2047,Energy (Atwater General Factors),KCAL,957,280.0
1,2048,Energy (Atwater Specific Factors),KCAL,958,290.0
2,1001,Solids,G,201,200.0


# Analiza poszczególnych plików
## Analiza na podstawie dokumentacji USDA + analizy rekordów
#### 1. food.csv
Główna tabela „spisu treści” USDA. Każdy produkt w FoodData Central ma tutaj rekord.

Zawiera:

- fdc_id – główne ID produktu

- data_type – typ danych (branded_food, foundation_food)

- description – nazwa produktu

- publication_date – data publikacji/aktualizacji

jakieś pola poboczne (food_category_id, market_country, microbe_data JSON)

Co realnie nas obchodzi:
`fdc_id`, `data_type`, `description`
Reszta to rządowe wypełniacze.

## 2. branded_food.csv

**Co to jest:**  
Rozszerzenie `food.csv` specyficzne dla produktów markowych.

**Zawiera m.in.:**
- `brand_owner` – firma (np. Campbell Soup)  
- `brand_name`, `subbrand_name`  
- `gtin_upc` – kod kreskowy  
- `ingredients` – skład (raw text)  
- `serving_size`, `serving_size_unit`  
- `household_serving_fulltext` – np. „1 Tbsp (15 ml)”  
- masa pól pobocznych (`preparation_state_code`, `trade_channel`, `discontinued_date`, itd.)

**Co realnie nas obchodzi:**  
`fdc_id`, `brand_owner`, `serving_size`, `serving_size_unit`, `household_serving_fulltext`, `gtin_upc`.

**Znaczenie:**  
Najważniejszy plik dla Dyeti.

---

## 3. food_nutrient.csv

**Co to jest:**  
Największy plik (ok. 1.4 GB). Każde powiązanie:  
`produkt (fdc_id)` → `składnik odżywczy (nutrient_id)` → `amount` per 100 g / ml.

**Zawiera:**  
- `fdc_id`  
- `nutrient_id`  
- `amount`  
- dodatkowe pola (`derivation_id`, `data_points`, min/max/median)

**Użycie:**  
Źródło makr i mikroelementów (białko, tłuszcz, żelazo, cynk, witamina C, potas itd.).

---

## 4. nutrient.csv

**Co to jest:**  
Słownik składników odżywczych.

**Zawiera:**  
- `id` (`nutrient_id`)  
- `name` – np. „Iron, Fe”, „Vitamin C, total ascorbic acid”  
- `unit_name` – mg, g, µg, kcal  
- `rank` – kolejność sortowania

**Użycie:**  
Pozwala interpretować `nutrient_id` z `food_nutrient.csv`.

---

## 5. measure_unit.csv

**Co to jest:**  
Słownik jednostek „household” (cup, tablespoon, teaspoon).

**Zawiera:**  
- `id` (`usdaUnitId`)  
- `name` – np. cup, tablespoon  
- czasem `abbreviation`

**Użycie:**  
Do mapowania jednostek w JSON-ie (np. `unit: { usdaUnitId: ... }`).

---

## 6. food_attribute.csv

**Co to jest:**  
Metadata o dodatkowych atrybutach produktów (np. „Ingredients”, „Nutrient Removed”).

**Realna wartość:**  
Zerowa dla naszych zastosowań.

---

## 7. food_attribute_type.csv

**Co to jest:**  
Słownik typów używanych w `food_attribute.csv`.

**Zawiera 3 rekordy:**  
- Update Log  
- Attribute  
- Common Name

**Użyteczność:**  
Można ignorować.

---

## 8. food_update_log_entry.csv

**Co to jest:**  
Historia aktualizacji produktu.

**Zawiera:**  
- `fdc_id`  
- `description`  
- `last_updated`

**Uwaga:**  
Informacje te są już w `food.publication_date`, więc zbędne.

---

## 9. microbe.csv

**Co to jest:**  
Dane mikrobiologiczne (bakterie, metody testowe).

**Zawiera:**  
- `foodId`  
- `microbe_code`  
- `method`  
- wartości liczbowe

**Użyteczność:**  
Dla naszej aplikacji brak.


## Wczytywanie tylko tych CSV, które nas interesują
czyli `food`, `branded_food`, `food_nutrient`, `nutrient` oraz `measure_unit`

In [1]:
import pandas as pd
from pathlib import Path

DATA_DIR = Path(r"C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\USDA_branded_foods")

print("Ładownie tylko potrzebnych CSV...")

# dtypes
food_dtypes = {
    "fdc_id": "int32",
    "data_type": "string",
    "description": "string",
}

branded_dtypes = {
    "fdc_id": "int32",
    "brand_owner": "string",
    "brand_name": "string",
    "subbrand_name": "string",
    "gtin_upc": "string",
    "ingredients": "string",
    "serving_size": "float32",
    "serving_size_unit": "string",
    "household_serving_fulltext": "string",
}

food_nutrient_dtypes = {
    "fdc_id": "int32",
    "nutrient_id": "int32",
    "amount": "float32",
}

nutrient_dtypes = {
    "id": "int32",
    "name": "string",
    "unit_name": "string",
}

# Umeasure_unit.csv ma TYLKO 'id' i 'name'
measure_unit_dtypes = {
    "id": "int32",
    "name": "string",
}

# FOOD
print(" - food.csv")
food = pd.read_csv(
    DATA_DIR / "food.csv",
    dtype=food_dtypes,
    usecols=["fdc_id", "data_type", "description", "publication_date"],
    low_memory=False,
)

# BRANDED FOOD
print(" - branded_food.csv")
branded = pd.read_csv(
    DATA_DIR / "branded_food.csv",
    dtype=branded_dtypes,
    usecols=[
        "fdc_id",
        "brand_owner", "brand_name",
        "serving_size", "serving_size_unit",
        "household_serving_fulltext",
        "gtin_upc"
    ],
    low_memory=False,
)

# FOOD NUTRIENT 
print(" - food_nutrient.csv")
food_nutrient = pd.read_csv(
    DATA_DIR / "food_nutrient.csv",
    dtype=food_nutrient_dtypes,
    usecols=["fdc_id", "nutrient_id", "amount"],
    low_memory=False,
)

# NUTRIENT
print(" - nutrient.csv")
nutrient = pd.read_csv(
    DATA_DIR / "nutrient.csv",
    dtype=nutrient_dtypes,
    usecols=["id", "name", "unit_name"],
    low_memory=False,
)

# MEASURE UNIT
print(" - measure_unit.csv")
measure_unit = pd.read_csv(
    DATA_DIR / "measure_unit.csv",
    dtype=measure_unit_dtypes,
    usecols=["id", "name"],
    low_memory=False,
)

# dodajemy sztucznie kolumnę abbreviation, żeby późniejszy kod, jakby czegoś oczekiwał, miał co czytać
measure_unit["abbreviation"] = pd.NA

print("wczytano tylko potrzebne dataset'y)")

# szybki check
print("\nshapes:")
print("  food:", food.shape)
print("  branded:", branded.shape)
print("  food_nutrient:", food_nutrient.shape)
print("  nutrient:", nutrient.shape)
print("  measure_unit:", measure_unit.shape)


Ładownie tylko potrzebnych CSV...
 - food.csv
 - branded_food.csv
 - food_nutrient.csv
 - nutrient.csv
 - measure_unit.csv
wczytano tylko potrzebne dataset'y)

shapes:
  food: (1977398, 4)
  branded: (1977398, 7)
  food_nutrient: (25652681, 3)
  nutrient: (477, 3)
  measure_unit: (122, 3)


## Ładownie danych do Data Frame'a

### 1 - wybór produktów *branded* i dołączenie podstawowych pól

- Kolumna `data_type` w `food.csv` jest normalizowana (zamiana na małe litery, usunięcie zbędnych spacji), a następnie wybierane są wyłącznie rekordy typu `branded_food` lub `branded`.
- Z przefiltrowanej tabeli pozostawiane są kolumny `fdc_id`, `description` oraz `publication_date`.
- Do tych rekordów dołączane są dane z `branded_food.csv` (po `fdc_id`), w szczególności:
  - `brand_owner`, `brand_name`
  - `serving_size`, `serving_size_unit`
  - `household_serving_fulltext`
  - `gtin_upc`
- Wynik jest zapisywany w zmiennej `food_b` jako bazowa tabela produktów markowych z identyfikatorem USDA, nazwą, datą publikacji oraz informacjami o producencie, porcji i kodzie kreskowym.


In [2]:
import pandas as pd
import re

# wybieramy tylko branded_food z food.csv
food_fix = food.copy()
food_fix["data_type_l"] = food_fix["data_type"].str.lower().str.strip()

food_b = (
    food_fix.loc[
        food_fix["data_type_l"].isin(["branded_food", "branded"]),
        ["fdc_id", "description", "publication_date"]
    ]
    .copy()
)

print("Branded-only (po filtrze):", food_b.shape)

# dokładamy pola z branded_food.csv, w tym brand_owner / brand_name / gtin_upc
food_b = food_b.merge(
    branded[[
        "fdc_id",
        "brand_owner",
        "brand_name",
        "serving_size", "serving_size_unit",
        "household_serving_fulltext",
        "gtin_upc",
    ]],
    on="fdc_id",
    how="left",
)

print("Branded-only po merge z branded:", food_b.shape)
display(food_b.head(5))


Branded-only (po filtrze): (1977398, 3)
Branded-only po merge z branded: (1977398, 9)


,fdc_id,description,publication_date,brand_owner,brand_name,serving_size,serving_size_unit,household_serving_fulltext,gtin_upc
0,1105904,WESSON Vegetable Oil 1 GAL,2020-11-13,Richardson Oilseed Products (US) Limited,<NA>,15.0,ml,1 Tbsp (15 ml),00027000612323
1,1105905,SWANSON BROTH BEEF,2020-11-13,CAMPBELL SOUP COMPANY,<NA>,240.0,ml,Amount per serving,00051000198808
2,1105906,CAMPBELL'S SLOW KETTLE SOUP CLAM CHOWDER,2020-11-13,CAMPBELL SOUP COMPANY,<NA>,440.0,g,PER CONTAINER,00051000213273
3,1105907,CAMPBELL'S SLOW KETTLE SOUP CHEESE BROCCOLI,2020-11-13,CAMPBELL SOUP COMPANY,<NA>,440.0,g,PER CONTAINER,00051000213303
4,1105908,SWANSON BROTH CHICKEN,2020-11-13,CAMPBELL SOUP COMPANY,<NA>,240.0,ml,Amount per Serving,00051000224637


### makra per 100 g i merge z produktami

- wybieramy z `food_nutrient.csv` tylko makroskładniki: kalorie, białko, węgle, tłuszcz
- robimy pivot tak, żeby dla każdego `fdc_id` była jedna linia z kolumnami:
  - `kcal100g`
  - `protein100g`
  - `carbs100g`
  - `fat100g`
- łączymy te makra z tabelą `food_b` tak, aby każdy produkt branded miał komplet danych żywieniowych na 100 g
- wynik to `df_branded_macros`, czyli tabela produktów z podstawowymi informacjami i makrami


In [3]:
# mapowanie nutrient_id -> nazwa kolumny 
NUTRIENT_MAP = {
    1008: "kcal100g",     # kcal
    1003: "protein100g",  # białko
    1005: "carbs100g",    # węglowodany
    1004: "fat100g",      # tłuszcz
}

# bierzemy tylko makra
fn_macros = food_nutrient[food_nutrient["nutrient_id"].isin(NUTRIENT_MAP.keys())]

# pivot: jedna linia per fdc_id, kolumny = kcal100g, protein100g, ...
fn_piv = (
    fn_macros.pivot_table(
        index="fdc_id",
        columns="nutrient_id",
        values="amount",
        aggfunc="first",
    )
    .rename(columns=NUTRIENT_MAP)
    .reset_index()
)

print("fn_piv (makra):", fn_piv.shape)
display(fn_piv.head(5))

# merge: produkt + makra
df_branded_macros = (
    food_b.merge(fn_piv, on="fdc_id", how="left")
          .sort_values("publication_date")
          .reset_index(drop=True)
)

print("df_branded_macros:", df_branded_macros.shape)
display(df_branded_macros.head(5))

print("\nNon-null counts (makra):")
print(df_branded_macros[["protein100g","carbs100g","fat100g","kcal100g"]].notna().sum())


fn_piv (makra): (1869938, 5)


nutrient_id,fdc_id,protein100g,fat100g,carbs100g,kcal100g
0,344604,0.810000,0.41,4.07,24.0
1,344605,0.810000,0.41,4.07,24.0
2,344606,23.209999,2.68,0.00,NaN
3,344607,23.209999,2.68,0.00,NaN
4,344608,18.750000,15.18,0.00,NaN


df_branded_macros: (1977398, 13)


,fdc_id,description,publication_date,brand_owner,brand_name,serving_size,serving_size_unit,household_serving_fulltext,gtin_upc,protein100g,fat100g,carbs100g,kcal100g
0,554495,HAMBURGERS,2019-04-01,White Castle System Inc.,<NA>,90.0,g,"1 PACKAGE, 2 SANDWICHES",082988000777,13.33,14.44,30.00,300.0
1,506065,ANTIPASTO DELIGHT,2019-04-01,"George DeLallo Co., Inc.",<NA>,113.0,g,1 PKG,072368106953,0.00,3.10,7.08,53.0
2,506066,UNCURED MAPLE SMOKED HAM,2019-04-01,WATERHILL NATURALS,<NA>,56.0,g,2 SLICES,052294066194,16.07,3.57,7.14,107.0
3,506067,GRANULATED ONION,2019-04-01,The C. F. Sauer Co.,<NA>,1.0,g,0.25 tsp,052500022211,0.00,0.00,100.00,500.0
4,506068,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA",2019-04-01,Lunds Inc,<NA>,42.0,g,0.25 cup,072431043857,14.29,5.95,71.43,381.0



Non-null counts (makra):
protein100g    1861596
carbs100g      1852671
fat100g        1839806
kcal100g       1840500
dtype: int64


### Jednostki porcji dla Branded Foods

W katalogu *Branded Foods* ( w przeciwieńswie do *foundational food* ) USDA **nie ma tabeli `food_portion`**, więc brak jest oficjalnych `gram_weight` dla jednostek typu *slice*, *cup*, *piece* itd. Plik `measure_unit.csv` jest tylko słownikiem nazw jednostek (ID -> nazwa), bez informacji o masie. Dlatego trzeba będzie być kreatywnym z parsowaniem `household_serving_fulltext`

W tej komórce:

- dla każdego produktu z `df_branded_macros` buduje jednostkę porcji:
  - `unit_usdaUnitId`
  - `unit_name`
  - `unit_symbol`
  - `gramsPerUnit`
- gdy `serving_size_unit` = `"g"`:
  - odczytywana jest masa porcji podana przez USDA (`serving_size`, np. 30 g)
  - parser sprawdza, czy `household_serving_fulltext` zaczyna się od wzorca typu `"X unit"` (np. `"2 pieces"`, `"1 slice"`)
  - jeśli uda się dopasować liczbę i nazwę jednostki:
      - liczba jednostek = X (np. 2)
      - `gramsPerUnit` = (`serving_size` w gramach) / X  
        przykład: 30 g / 2 = 15 g na 1 piece
      - jeśli nazwa jednostki (np. `"piece"`, `"slice"`, `"cup"`) występuje w `measure_unit.csv`, to ustawiany jest `unit_usdaUnitId`
      - jeśli jednostka nie istnieje w `measure_unit.csv`, `unit_usdaUnitId` pozostaje `None`, ale `unit_name` i `gramsPerUnit` są zachowane

- gdy `serving_size_unit` = `"ml"` lub dane są nieparsowalne / podejrzane:
  - stosuje bezpieczny fallback: `unit_name = "gram"`, `unit_symbol = "g"`, `gramsPerUnit = 1.0`, `unit_usdaUnitId = None`

Uzupełnianie brakujących `unit_usdaUnitId` ( zostanie to zrobione analogicznie jak w analizie *foundational foods* ) jest robione później, na gotowym DataFrame.


In [6]:
import re

# przygotowanie measure_unit (tylko name/id)
mu = measure_unit.copy()
mu["name_l"] = mu["name"].str.lower().str.strip()

def find_unit_id(name: str):
    if not isinstance(name, str):
        return None
    s = name.strip().lower()
    hit = mu[mu["name_l"] == s]
    if len(hit):
        return int(hit.iloc[0]["id"])
    return None

# aliasy do parsowania jednostek z household_serving_fulltext
UNIT_ALIASES = {
    "tsp": "teaspoon",
    "teaspoons": "teaspoon",
    "tbsp": "tablespoon",
    "tablespoons": "tablespoon",
    "cup": "cup",
    "cups": "cup",
    "slice": "slice",
    "slices": "slice",
    "piece": "piece",
    "pieces": "piece",
    "serving": "serving",
    "servings": "serving",
    "container": "container",
    "containers": "container",
    "pkg": "package",
    "package": "package",
    "packages": "package",
    "g": "gram",
}

NUM_UNIT_RE = re.compile(r"^\s*(\d+(?:\.\d+)?)\s*([a-zA-Z]+)")

def parse_household(txt: str):
    
    #Zwraca (qty, unit_norm) jeśli tekst zaczyna się od 'X unit'
    #(np. '2 pieces', '1 slice'). W przeciwnym razie None.
    
    if not isinstance(txt, str):
        return None
    m = NUM_UNIT_RE.match(txt)
    if not m:
        return None
    qty = float(m.group(1))
    unit_raw = m.group(2).lower()
    unit_norm = UNIT_ALIASES.get(unit_raw, unit_raw)
    return qty, unit_norm


def build_unit_row(row):
    serving_size = row.serving_size
    su = row.serving_size_unit
    serving_unit = (su or "").strip().lower() if isinstance(su, str) else ""
    household_txt = row.household_serving_fulltext

    # domyślny fallback: gram / 1 g, ale unit_usdaUnitId = None
    unit_id = None
    unit_name = "gram"
    unit_symbol = "g"
    grams_per_unit = 1.0

    # przypadek ml -> twardy fallback
    if serving_unit in ("ml", "milliliter"):
        return pd.Series({
            "unit_usdaUnitId": unit_id,
            "unit_name": unit_name,
            "unit_symbol": unit_symbol,
            "gramsPerUnit": grams_per_unit,
        })

    # przypadek gram -> próbujemy sparsować household_serving_fulltext
    if serving_unit in ("g", "gram"):
        parsed = parse_household(household_txt)
        if parsed and pd.notna(serving_size) and serving_size > 0:
            qty, unit_txt = parsed
            if qty > 0:
                uid = find_unit_id(unit_txt)
                if uid is not None:
                    unit_id = uid   # USDA zna tę jednostkę
                else:
                    unit_id = None  # USDA nie zna, ale nazwa i tak zostaje

                unit_name = unit_txt
                unit_symbol = unit_txt
                grams_per_unit = float(serving_size) / float(qty)

    # reszta → fallback gram / 1 g

    return pd.Series({
        "unit_usdaUnitId": unit_id,
        "unit_name": unit_name,
        "unit_symbol": unit_symbol,
        "gramsPerUnit": grams_per_unit,
    })


# zastosowanie do całego DF z makrami
unit_df = df_branded_macros.apply(build_unit_row, axis=1)

# tu dokładamy też brand_owner / brand_name
df_products = pd.concat(
    [
        df_branded_macros[[
            "fdc_id",
            "description",
            "brand_owner",
            "brand_name",
            "kcal100g",
            "protein100g",
            "carbs100g",
            "fat100g",
        ]],
        unit_df,
    ],
    axis=1,
)

df_products = df_products.rename(columns={
    "fdc_id": "usdaId",
    "description": "name",
})

# tutaj powstaje nazwa "nazwa_produktu - marka"
def build_display_name(row):
    base = row["name"]
    brand = row["brand_name"]
    owner = row["brand_owner"]

    if isinstance(brand, str) and brand.strip():
        return f"{base} - {brand.strip()}"
    if isinstance(owner, str) and owner.strip():
        return f"{base} - {owner.strip()}"
    return base

df_products["displayName"] = df_products.apply(build_display_name, axis=1)

print("df_products:", df_products.shape)
display(df_products[["usdaId", "name", "brand_name", "brand_owner", "displayName"]].head(5))


df_products: (1977398, 13)


,usdaId,name,brand_name,brand_owner,displayName
0,554495,HAMBURGERS,<NA>,White Castle System Inc.,HAMBURGERS - White Castle System Inc.
1,506065,ANTIPASTO DELIGHT,<NA>,"George DeLallo Co., Inc.","ANTIPASTO DELIGHT - George DeLallo Co., Inc."
2,506066,UNCURED MAPLE SMOKED HAM,<NA>,WATERHILL NATURALS,UNCURED MAPLE SMOKED HAM - WATERHILL NATURALS
3,506067,GRANULATED ONION,<NA>,The C. F. Sauer Co.,GRANULATED ONION - The C. F. Sauer Co.
4,506068,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA",<NA>,Lunds Inc,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA - Lunds..."


Finalna nazwa produktu powstała poprzez połączenie `product_name` + "`-`" +  ( `brand_name` lub `brand_owner`).
Bez tego, finalnie mielibyśmy 50 produktów o nazwie np po prostu "chips", teraz wiemy, czy to są chipsy marki Lays albo Crunchips.  
Pełen DF:

In [7]:
display(df_products.head(5))

,usdaId,name,brand_owner,brand_name,kcal100g,protein100g,carbs100g,fat100g,unit_usdaUnitId,unit_name,unit_symbol,gramsPerUnit,displayName
0,554495,HAMBURGERS,White Castle System Inc.,<NA>,300.0,13.33,30.00,14.44,1039.0,package,package,90.0,HAMBURGERS - White Castle System Inc.
1,506065,ANTIPASTO DELIGHT,"George DeLallo Co., Inc.",<NA>,53.0,0.00,7.08,3.10,1039.0,package,package,113.0,"ANTIPASTO DELIGHT - George DeLallo Co., Inc."
2,506066,UNCURED MAPLE SMOKED HAM,WATERHILL NATURALS,<NA>,107.0,16.07,7.14,3.57,1050.0,slice,slice,28.0,UNCURED MAPLE SMOKED HAM - WATERHILL NATURALS
3,506067,GRANULATED ONION,The C. F. Sauer Co.,<NA>,500.0,0.00,100.00,0.00,1002.0,teaspoon,teaspoon,4.0,GRANULATED ONION - The C. F. Sauer Co.
4,506068,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA",Lunds Inc,<NA>,381.0,14.29,71.43,5.95,1000.0,cup,cup,168.0,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA - Lunds..."


# "Surowy" Data Frame jest gotowy
#### teraz przechodzimy do "sanity check-ów", wypełniania braków, usuwania bezużytecznych rekordów itp Analogicznie jak w analizie danych z *foundational foods*

### 1. Sprawdzenie, czy gdzieś nie ma nulla w makrach

In [7]:
# sprawdzamy null'e w 3 głównych makrach
null_mask = (
    df_products["protein100g"].isna() |
    df_products["carbs100g"].isna() |
    df_products["fat100g"].isna()
)

missing_macros = df_products[null_mask]

print("Liczba produktów z brakującymi makrami:", missing_macros.shape[0])
display(missing_macros.head(10))


Liczba produktów z brakującymi makrami: 154377


,usdaId,name,brand_owner,brand_name,kcal100g,protein100g,carbs100g,fat100g,unit_usdaUnitId,unit_name,unit_symbol,gramsPerUnit,displayName
213,505993,SEAFOOD HOUSE ALL PURPOSE RUB,ADAMS RESERVE,<NA>,0.0,NaN,NaN,0.00,1002.0,teaspoon,teaspoon,2.000000,SEAFOOD HOUSE ALL PURPOSE RUB - ADAMS RESERVE
220,506000,SEASONING,ADAMS RESERVE,<NA>,0.0,NaN,NaN,0.00,1001.0,tablespoon,tablespoon,3.600000,SEASONING - ADAMS RESERVE
383,506451,"ECHIRE, BEURRE DE BARATTE D' EXCELLENCE",ECHIRE,<NA>,723.0,0.50,NaN,80.00,NaN,grm,grm,1.000000,"ECHIRE, BEURRE DE BARATTE D' EXCELLENCE - ECHIRE"
622,505484,SUGAR COOKIE MIX ROLL & CUT,PAMELA'S,<NA>,364.0,0.00,95.449997,NaN,NaN,th,th,372.881356,SUGAR COOKIE MIX ROLL & CUT - PAMELA'S
623,505483,CHOCOLATE CHUNKCOOKIE MIX,PAMELA'S,<NA>,476.0,4.76,76.190002,NaN,NaN,th,th,375.000000,CHOCOLATE CHUNKCOOKIE MIX - PAMELA'S
864,505794,SANDWICH BREAD MIX,BREADY,<NA>,375.0,NaN,75.000000,2.08,1039.0,package,package,436.363636,SANDWICH BREAD MIX - BREADY
1206,507301,PREMIUM MASHED POTATOES,"Idahoan Foods, LLC",<NA>,385.0,7.69,69.230003,NaN,1000.0,cup,cup,52.000000,"PREMIUM MASHED POTATOES - Idahoan Foods, LLC"
1903,506489,MICHELADA FILL WITH ICE COLD BEER,DON CHELADA,<NA>,125.0,3.33,4.170000,NaN,NaN,gram,g,1.000000,MICHELADA FILL WITH ICE COLD BEER - DON CHELADA
2004,506966,WHOLE BLACK OLIVES,AramcO Imports Inc.,<NA>,112.0,NaN,12.500000,12.50,NaN,grm,grm,1.000000,WHOLE BLACK OLIVES - AramcO Imports Inc.
2121,505308,DRIED CRANBERRIES,HARMOUS,<NA>,354.0,NaN,84.000000,2.00,NaN,grm,grm,1.000000,DRIED CRANBERRIES - HARMOUS


#### Jak widać, mamy ponad 150 tysięcy takich rekordów, niestety są one bezużyteczne dla dYeti, więc je usuwamy.

In [8]:
# usuwamy rekordy z brakującymi makrami (protein / carbs / fat)

mask_missing_macros = df_products[["protein100g", "carbs100g", "fat100g"]].isna().any(axis=1)

n_total = df_products.shape[0]
n_bad = int(mask_missing_macros.sum())
n_good = n_total - n_bad

print(f"Łącznie rekordów: {n_total}")
print(f"Z brakującymi makrami (protein/carbs/fat): {n_bad}")
print(f"Pozostaje rekordów po odfiltrowaniu: {n_good}")

# odfiltrowanie
df_products = df_products[~mask_missing_macros].reset_index(drop=True)

print("\nPo czyszczeniu:")
print(df_products.shape)
print(df_products[["protein100g","carbs100g","fat100g"]].notna().sum())


Łącznie rekordów: 1977398
Z brakującymi makrami (protein/carbs/fat): 154377
Pozostaje rekordów po odfiltrowaniu: 1823021

Po czyszczeniu:
(1823021, 13)
protein100g    1823021
carbs100g      1823021
fat100g        1823021
dtype: int64


### 2. Szukanie rekordów z ujemnymi makrami

In [9]:
# sprawdzamy, czy gdzieś makra mają wartości ujemne

invalid_rows = df_products[
    (df_products["protein100g"] < 0)
    | (df_products["fat100g"] < 0)
    | (df_products["carbs100g"] < 0)
]

if invalid_rows.empty:
    print("Brak rekordów z ujemnymi wartościami białka, tłuszczu lub węglowodanów.")
else:
    print(f"Wykryto {len(invalid_rows)} rekordów z błędnymi wartościami:")
    display(
        invalid_rows[[
            "usdaId", 
            "name", 
            "protein100g", 
            "fat100g", 
            "carbs100g"
        ]].head(50)
    )


Brak rekordów z ujemnymi wartościami białka, tłuszczu lub węglowodanów.


#### Miłe zaskoczenie

### 3. Czy produkty się mieszczą w [ 0, 900 ] kcal/100gram ( teorytyczny limit )

In [10]:
# sanity check: wartości ekstremalne kcal / 100 g

max_row = df_products.loc[df_products["kcal100g"].idxmax()]
min_row = df_products.loc[df_products["kcal100g"].idxmin()]

print("Najwięcej kcal/100 g:")
print(f'  {max_row["name"]} (usdaId={max_row["usdaId"]}) — {max_row["kcal100g"]} kcal/100g')

print("\nNajmniej kcal/100 g:")
print(f'  {min_row["name"]} (usdaId={min_row["usdaId"]}) — {min_row["kcal100g"]} kcal/100g')


Najwięcej kcal/100 g:
  FRENCH VANILLA MONK FRUIT ORGANIC SWEETENER CONCENTRATE, FRENCH VANILLA (usdaId=2281961) — 151515.0 kcal/100g

Najmniej kcal/100 g:
  HIZIKI SEA VEGETABLE HAND HARVESTED WILD (usdaId=506082) — 0.0 kcal/100g


Jak widać na przykładzie *FRENCH VANILLA MONK FRUIT ORGANIC SWEETENER CONCENTRATE, FRENCH VANILLA (usdaId=2281961) — 151515.0 kcal/100g*, istnieją tutaj rekordy o absurdalnych wartościach  ( ponad 150 tysięcy kalorii na 100 gram produktu ). Mogą być tutaj "błedne" rekordy, gdzie wina stoi po stronie producenta. Jednakże takich absurdalnych produktów też się trzeba pozbyć. Za cutoff "absurdu" przyjmujemy 950 kcal / 100 gram ( 50 kcal tolerancji).

In [13]:
suspicious_kcal = df_products[df_products["kcal100g"] > 950]

print(f"Liczba podejrzanych produktów (>950 kcal/100g): {suspicious_kcal.shape[0]}")
display(suspicious_kcal.head(10))


Liczba podejrzanych produktów (>950 kcal/100g): 4183


,usdaId,name,brand_owner,brand_name,kcal100g,protein100g,carbs100g,fat100g,unit_usdaUnitId,unit_name,unit_symbol,gramsPerUnit,displayName
4427,504359,SWEET POTATO WALNUTS,So Nuts and Confections LLC,<NA>,1000.0,14.29,100.00,64.290001,NaN,onz,onz,28.0,SWEET POTATO WALNUTS - So Nuts and Confections...
8076,509393,CHEDDAR PIMENTO CHEESE,MRS. MARSHALL'S,<NA>,1179.0,14.29,0.00,42.860001,1001.0,tablespoon,tablespoon,14.0,CHEDDAR PIMENTO CHEESE - MRS. MARSHALL'S
9161,496757,OLIVE OIL,"La Tourangelle, Inc",<NA>,1000.0,0.00,0.00,100.000000,NaN,gram,g,1.0,"OLIVE OIL - La Tourangelle, Inc"
9931,496393,PURE SESAME OIL,ASSI,<NA>,1113.0,26.67,26.67,100.000000,NaN,gram,g,1.0,PURE SESAME OIL - ASSI
10197,497856,"LA TOURANGELLE, ROASTED WALNUT SPRAY OIL","La Tourangelle, Inc",<NA>,1000.0,0.00,0.00,100.000000,NaN,gram,g,1.0,"LA TOURANGELLE, ROASTED WALNUT SPRAY OIL - La ..."
12041,493632,VEGAN NON-AEROSOL SPRAY OIL,"The Hain Celestial Group, Inc.",<NA>,1000.0,0.00,0.00,100.000000,NaN,gram,g,1.0,VEGAN NON-AEROSOL SPRAY OIL - The Hain Celesti...
12043,493631,CANOLA & OLIVE SPRAY OIL,"The Hain Celestial Group, Inc.",<NA>,1000.0,0.00,0.00,100.000000,NaN,gram,g,1.0,CANOLA & OLIVE SPRAY OIL - The Hain Celestial ...
12742,495777,100% PURE SESAME OIL,Rhee Bros. Inc.,<NA>,1113.0,0.00,0.00,86.669998,NaN,gram,g,1.0,100% PURE SESAME OIL - Rhee Bros. Inc.
13861,501554,ROASTED PISTACHIO SPRAY OIL,LA TOURANGELLE,<NA>,1000.0,0.00,0.00,100.000000,NaN,gram,g,1.0,ROASTED PISTACHIO SPRAY OIL - LA TOURANGELLE
28884,515952,EDIBLE DECOR PUMPKINS,Xcell International Corp,<NA>,1667.0,0.00,100.00,0.000000,1001.0,tablespoon,tablespoon,6.0,EDIBLE DECOR PUMPKINS - Xcell International Corp


#### Usuwanie "suspicuous" produktów

In [14]:
print("Liczba rekordów przed czyszczeniem:", df_products.shape[0])

high_kcal = df_products[df_products["kcal100g"] > 950]
print(f"Liczba produktów >950 kcal/100g: {high_kcal.shape[0]}")

df_products = df_products[df_products["kcal100g"] <= 950].reset_index(drop=True)

print("Liczba rekordów po czyszczeniu:", df_products.shape[0])


Liczba rekordów przed czyszczeniem: 1823021
Liczba produktów >950 kcal/100g: 4183
Liczba rekordów po czyszczeniu: 1793667


### 4. Czy produkty mieszczą się w aproksymacji kcal = carbs * 4 + protein * 4 + fats * 9
Za "suspicious" produkty uznajemy teraz te, które odbbiegają od tej formuły o ponad 50 kcal

In [16]:
print("Sprawdzam spójność kcal z makrami")

# kcal teoretyczne
df_products["kcal_calc"] = (
    4 * df_products["protein100g"] +
    4 * df_products["carbs100g"] +
    9 * df_products["fat100g"]
)

# różnica absolutna
df_products["kcal_diff"] = (df_products["kcal_calc"] - df_products["kcal100g"]).abs()

# podejrzane, gdy różnica > 50 kcal na 100 g
suspect_energy = df_products[df_products["kcal_diff"] > 50]

print(f"Liczba produktów z niespójną energią (>50 kcal różnicy): {suspect_energy.shape[0]}")

display(
    suspect_energy[
        ["usdaId", "displayName", "kcal100g", "kcal_calc", "kcal_diff", 
         "protein100g", "carbs100g", "fat100g"]
    ].head(10)
)


Sprawdzam spójność kcal z makrami
Liczba produktów z niespójną energią (>50 kcal różnicy): 96874


,usdaId,displayName,kcal100g,kcal_calc,kcal_diff,protein100g,carbs100g,fat100g
3,506067,GRANULATED ONION - The C. F. Sauer Co.,500.0,400.000000,100.000000,0.00,100.000000,0.0
50,506042,MEDLEY EXOTIC DRIED MUSHROOM - World Variety P...,357.0,285.720001,71.279999,14.29,57.139999,0.0
174,505920,"CAKE CUPS - Safeway, Inc.",400.0,320.000000,80.000000,0.00,80.000000,0.0
217,505998,MEMPHIS BARBECUE RUB - ADAMS RESERVE,500.0,400.000000,100.000000,0.00,100.000000,0.0
218,505999,THAI THAI SEASONING - Adams Extract Co.,500.0,400.000000,100.000000,100.00,0.000000,0.0
244,505990,TEXAS STYLE SMOKED RIBS - Adams Extract Co.,500.0,0.000000,500.000000,0.00,0.000000,0.0
292,506372,"100% LEMON JUICE - George DeLallo Co., Inc.",0.0,80.000000,80.000000,0.00,20.000000,0.0
307,506384,"GARLIC IN OIL - Transnational Foods, Inc.",633.0,170.000000,463.000000,6.67,13.330000,10.0
331,506341,HIMALAYAN BLACK PEPPER GRINDER - WBM LLC,500.0,400.000000,100.000000,0.00,100.000000,0.0
340,506359,"YUENGLING, BROWN SUGAR BBQ DRY RUB - YUENGLING",385.0,307.679993,77.320007,0.00,76.919998,0.0


#### Jak widać, takich poroduktów jest dośc dużo - prawie 100 tysięcy
Jednakże jest to niewielka część całego dataset'a, a odchyelnie ponad 50 kcal jest już podejrzane i może sugerować błąd po stronie producenta, więc lepiej się tych rekordów pozbyć. 

In [17]:
print("Liczba rekordów przed czyszczeniem:", df_products.shape[0])

bad_energy_mask = df_products["kcal_diff"] > 50
bad_count = bad_energy_mask.sum()

print(f"Liczba produktów z niespójnością energetyczną >50 kcal/100g: {bad_count}")

df_products = df_products.loc[~bad_energy_mask].reset_index(drop=True)

print("Liczba rekordów po czyszczeniu:", df_products.shape[0])


Liczba rekordów przed czyszczeniem: 1793667
Liczba produktów z niespójnością energetyczną >50 kcal/100g: 96874
Liczba rekordów po czyszczeniu: 1696793


## 5. Normalizacja unit
Rekordy gdzie Unit ID ( nie unit name) jest null, czyli te, które nie mają "oficjalnego" unit ID tylko porcja jest podana przez producenta, tutaj jest, żeby powiedzieć grzecznie, "wątpliwej jakości" danych na temat porcji.

In [21]:
mask_no_usda_unit = df_products["unit_usdaUnitId"].isna()

count_no_usda_unit = mask_no_usda_unit.sum()
count_total = df_products.shape[0]

print(f"Liczba produktów bez USDA unit ID: {count_no_usda_unit}")
print(f"Procent: {count_no_usda_unit / count_total * 100:.2f}%")


Liczba produktów bez USDA unit ID: 1005082
Procent: 59.23%


Jak widać, jest to prawie 60% produktów

In [22]:
display(
    df_products.loc[mask_no_usda_unit, ["usdaId", "name", "unit_name", "unit_symbol", "gramsPerUnit"]]
             .head(20)
)

,usdaId,name,unit_name,unit_symbol,gramsPerUnit
8,506073,APPLE SODA,gram,g,1.000000
9,506074,"GO VEGGIE!, PASTEURIZED PROCESS CHEESE FOOD AL...",onz,onz,28.000000
10,506075,"GO VEGGIE!, PASTEURIZED PROCESS CHEESE FOOD AL...",onz,onz,28.000000
11,506076,GLAZED LEMON PIE,onz,onz,28.444444
21,506095,"KEEBLER, FUDGE GRAHAMS CRACKERS",pouch,pouch,20.000000
22,506094,"NABISCO, RITZ, CRACKERFULS FILLED CRACKERS, PE...",pack,pack,28.000000
23,506064,FRESH GROUND TURKEY,onz,onz,28.000000
24,506093,CLASSIC RED HAND CRAFTED SODA,gram,g,1.000000
25,506091,"MINUTE MAID, PINK LEMONADE JUICE",gram,g,1.000000
26,506090,"DR PEPPER, SODA",gram,g,1.000000


### Dla przypomnienia - Jak powstaje `unit_name` w Branded Foods?

Branded Foods nie mają tabeli `food_portion`, więc USDA nie podaje oficjalnych jednostek ani gramatur porcji.  
Dlatego `unit_name` pochodzi z dwóch źródeł:

1. **Parsowanie `household_serving_fulltext`**  
   Jeśli tekst wygląda jak `X unit` (np. `2 slices`, `1 piece`, `3 tbsp`), parser wyciąga liczbę i nazwę jednostki, a następnie liczy:  
   `gramsPerUnit = serving_size / X`.

2. **Fallback**  
   Jeśli parsowanie się nie powiodło lub jednostka jest bez sensu, ustawiane jest:  
   `unit_name = "gram"`, `unit_symbol = "g"`, `gramsPerUnit = 1.0`, `unit_usdaUnitId = None`.


### Nadawanie sztucznych `usdaUnitId` dla niestandardowych jednostek

Branded Foods posiadają wiele jednostek, które nie występują w oficjalnej tabeli `measure_unit`
(USDA unit IDs mieszczą się w zakresie 1000–2000).  
Dla takich przypadków tworzymy własne, stabilne identyfikatory:

- `gram` zachowuje wcześniej przyjęty placeholder ( z poprzedniej obróbki danych *foundational foods`) `usdaUnitId = 2048`,
- każda inna niestandardowa jednostka dostaje nowe ID zaczynające się od **3000**,
- numery są nadawane deterministycznie, według sortowanej listy unikalnych `unit_name`.

Dzięki temu unikamy kolizji z USDA i możemy później łatwo zarządzać dodatkowymi jednostkami.

Na początek zajmiemy się tą specjalną, niepoliczalną jednostką "gram". Wypiszemy wszystkie rekordy które jako `unit_name` mają jakąś jednostke wagi, także "amerykańską" ( np uncja ) oraz wszystkie te, co się zaczynają na `g`

In [23]:
import re

all_units = sorted(set(df_products["unit_name"].dropna()))

# wzorce jednostek wagi
weight_regex = re.compile(
    r"^(g|gm|gr|gram|grm|gms|ounce|oz|lb|lbs|pound|pounds|kg|kilogram)", 
    flags=re.IGNORECASE
)

weight_like_units = [u for u in all_units if weight_regex.match(u)]

print("Jednostki wyglądające na wagę:")
for u in weight_like_units:
    print(" -", u)

print("\nŁącznie:", len(weight_like_units))


Jednostki wyglądające na wagę:
 - galette
 - galettes
 - gar
 - garlic
 - ge
 - gel
 - gems
 - genoa
 - german
 - get
 - gf
 - gherkins
 - ghost
 - ghosts
 - gingerbread
 - glass
 - glazed
 - gm
 - go
 - gordita
 - gr
 - graham
 - grahams
 - gram
 - grams
 - granola
 - grape
 - grapefruit
 - gratin
 - gratins
 - grilled
 - griller
 - grinds
 - grm
 - gum
 - gumball
 - gummies
 - gummy
 - gushers
 - gyro
 - lb
 - lbs
 - ounce
 - ounces
 - oz
 - oza
 - ozs

Łącznie: 47


#### Normalizacja pseudojednostek wagi

Branded Foods zawiera wiele *"dziwnych"* wariantów typu *grm*, *gm*, *ozs*, *oza*, *ounce*, które w praktyce oznaczają to samo.  
Nie mają one wiarygodnych gramWeight, więc wszystkie normalizujemy do jednej jednostki: **gram (ID 2048, 1 g)**.  

In [25]:
weight_like_units = {
    "gm", "gr", "grm", "gram", "grams",
    "oz", "oza", "ozs", "ounce", "ounces"
}

def normalize_weight_units(row):
    name = str(row["unit_name"]).lower().strip()

    if name in weight_like_units:
        return pd.Series({
            "unit_name": "gram",
            "unit_symbol": "g",
            "unit_usdaUnitId": 2048,
            "gramsPerUnit": 1.0
        })
    
    # wpp zwracamy oryginał
    return pd.Series({
        "unit_name": row["unit_name"],
        "unit_symbol": row["unit_symbol"],
        "unit_usdaUnitId": row["unit_usdaUnitId"],
        "gramsPerUnit": row["gramsPerUnit"]
    })

df_products[["unit_name", "unit_symbol", "unit_usdaUnitId", "gramsPerUnit"]] = \
    df_products.apply(normalize_weight_units, axis=1)

print("Normalizacja jednostek wagi wykonana.")


Normalizacja jednostek wagi wykonana.


Ogólnie można zauważyć, że pojawiają się `unit_name` co mają 1 lub 2 znaki długosci. Zakładamy, że unit typu `x` albo `th` nie wnosi nam żadnej wartościowej informacji, więc tak samo, sprrowadzimy je do fall-back unit: czyli grama

In [26]:
# Zamieniamy na string, żeby bezpiecznie liczyć długości
unit_names = df_products["unit_name"].astype("string")

mask_null = unit_names.isna()

mask_len1 = unit_names.str.len() == 1

mask_len2 = unit_names.str.len() == 2

print("Null unit_name:", mask_null.sum())
print("Length 1:", mask_len1.sum())
print("Length 2:", mask_len2.sum())

print("\nPrzykłady (len = 1):")
print(unit_names[mask_len1].dropna().unique()[:20])

print("\nPrzykłady (len = 2):")
print(unit_names[mask_len2].dropna().unique()[:20])


Null unit_name: 0
Length 1: 1138
Length 2: 5104

Przykłady (len = 1):
<StringArray>
['c', 'k', 't', 'i', 'o', 'x', 's', 'w', 'p', 'a', 'z', 'l', 'e', 'm']
Length: 14, dtype: string

Przykłady (len = 2):
<StringArray>
['of', 'th', 'pc', 'ml', 'fl', 'pf', 'pt', 'go', 'or', 'lb', 'in', 'sc', 'to',
 'fb', 'ct', 'pz', 'up', 'ea', 'pk', 'wt']
Length: 20, dtype: string


In [27]:
# Normalizacja jednostek o długości 0/1/2 znaków lub null
def normalize_short_units(row):
    name = row["unit_name"]

    if pd.isna(name) or len(str(name)) <= 2:
        return pd.Series({
            "unit_name": "gram",
            "unit_symbol": "g",
            "unit_usdaUnitId": 2048,
            "gramsPerUnit": 1.0
        })

    return pd.Series({
        "unit_name": row["unit_name"],
        "unit_symbol": row["unit_symbol"],
        "unit_usdaUnitId": row["unit_usdaUnitId"],
        "gramsPerUnit": row["gramsPerUnit"]
    })

df_products[["unit_name", "unit_symbol", "unit_usdaUnitId", "gramsPerUnit"]] = \
    df_products.apply(normalize_short_units, axis=1)

print("Normalizacja jednostek ≤2 znaki wykonana.")


Normalizacja jednostek ≤2 znaki wykonana.


Teraz dla pozostałych jednostek zrobimy "customowe" UsdaUnitID, zaczynające się od 3000, aby w razie czego możnaby było je łatwo odróżnić. Unit name i unit symbol będą identyczne. 

In [28]:
# Customowe usdaUnitId dla pozostałych jednostek (tam, gdzie unit_usdaUnitId jest null)

mask_custom = df_products["unit_usdaUnitId"].isna()

custom_unit_names = (
    df_products.loc[mask_custom, "unit_name"]
    .dropna()
    .sort_values()
    .unique()
)

print("Rekordów bez usdaUnitId:", mask_custom.sum())
print("Unikalnych unit_name w tej grupie:", len(custom_unit_names))

BASE_ID = 3000
name_to_id = {name: BASE_ID + idx for idx, name in enumerate(custom_unit_names)}

# przypisujemy nowe ID tylko tam, gdzie wcześniej było NaN
df_products.loc[mask_custom, "unit_usdaUnitId"] = (
    df_products.loc[mask_custom, "unit_name"].map(name_to_id)
)

# dla customowych jednostek symbol = nazwa
df_products.loc[mask_custom, "unit_symbol"] = df_products.loc[mask_custom, "unit_name"]

print("Przykładowe customowe jednostki:")
display(
    df_products.loc[mask_custom, ["unit_name", "unit_symbol", "unit_usdaUnitId"]]
              .drop_duplicates()
              .head(20)
)


Rekordów bez usdaUnitId: 392633
Unikalnych unit_name w tej grupie: 1267
Przykładowe customowe jednostki:


,unit_name,unit_symbol,unit_usdaUnitId
9,onz,onz,3687.0
21,pouch,pouch,3862.0
22,pack,pack,3702.0
29,cookies,cookies,3249.0
37,pepper,pepper,3763.0
57,frank,frank,3428.0
68,crackers,crackers,3269.0
117,sandwiches,sandwiches,3965.0
120,hot,hot,3494.0
252,cone,cone,3231.0


Niektóre mogą wydawać się dziwnie, ale obędzie się bez ręcznego sprawdzania +1000 rekordów.

# Gotowy Data Frame
W tej postaci DataFrame jest już praktycznie gotowy. Mamy w ten sposób prawie *1 700 000* produktów. Dodatkowo wszystko teraz zapiszemy do pliku csv.

In [33]:
print(df_products.shape)

display(df_products.head(20))

df_products.info()

(1696793, 15)


,usdaId,name,brand_owner,brand_name,kcal100g,protein100g,carbs100g,fat100g,unit_usdaUnitId,unit_name,unit_symbol,gramsPerUnit,displayName,kcal_calc,kcal_diff
0,554495,HAMBURGERS,White Castle System Inc.,<NA>,300.0,13.330000,30.000000,14.440000,1039.0,package,package,90.000000,HAMBURGERS - White Castle System Inc.,303.279999,3.279999
1,506065,ANTIPASTO DELIGHT,"George DeLallo Co., Inc.",<NA>,53.0,0.000000,7.080000,3.100000,1039.0,package,package,113.000000,"ANTIPASTO DELIGHT - George DeLallo Co., Inc.",56.220001,3.220001
2,506066,UNCURED MAPLE SMOKED HAM,WATERHILL NATURALS,<NA>,107.0,16.070000,7.140000,3.570000,1050.0,slice,slice,28.000000,UNCURED MAPLE SMOKED HAM - WATERHILL NATURALS,124.970001,17.970001
3,506068,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA",Lunds Inc,<NA>,381.0,14.290000,71.430000,5.950000,1000.0,cup,cup,168.000000,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA - Lunds...",396.429993,15.429993
4,506069,"SEEDS OF CHANGE, JALFREZI CURRY SAUCE, MEDIUM-HOT",Seeds Of Change Incorporated,<NA>,106.0,1.180000,10.590000,5.880000,1000.0,cup,cup,257.575758,"SEEDS OF CHANGE, JALFREZI CURRY SAUCE, MEDIUM-...",100.000000,6.000000
5,506070,"SEEDS OF CHANGE, MADRAS RED HOT CURRY SAUCE",Seeds Of Change Incorporated,<NA>,71.0,1.180000,9.410000,3.530000,1000.0,cup,cup,257.575758,"SEEDS OF CHANGE, MADRAS RED HOT CURRY SAUCE - ...",74.130005,3.130005
6,506071,"WILTON, CUPCAKE DECORATING KIT",Wilton Enterprises Inc,<NA>,375.0,0.000000,100.000000,0.000000,1002.0,teaspoon,teaspoon,4.000000,"WILTON, CUPCAKE DECORATING KIT - Wilton Enterp...",400.000000,25.000000
7,506072,"SANTA CRUZ, ORGANIC CRUNCHY DARK ROASTED PEANU...","SANTA CRUZ NATURAL, INC.",<NA>,656.0,21.879999,18.750000,50.000000,1001.0,tablespoon,tablespoon,16.000000,"SANTA CRUZ, ORGANIC CRUNCHY DARK ROASTED PEANU...",612.520020,43.479980
8,506073,APPLE SODA,MANZANITA SOL,<NA>,34.0,0.000000,9.300000,0.000000,2048.0,gram,g,1.000000,APPLE SODA - MANZANITA SOL,37.200001,3.200001
9,506074,"GO VEGGIE!, PASTEURIZED PROCESS CHEESE FOOD AL...","Galaxy Nutritional Foods, Inc.",<NA>,250.0,17.860001,3.570000,17.860001,3687.0,onz,onz,28.000000,"GO VEGGIE!, PASTEURIZED PROCESS CHEESE FOOD AL...",246.460007,3.539993


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1696793 entries, 0 to 1696792
Data columns (total 15 columns):
 #   Column           Dtype  
---  ------           -----  
 0   usdaId           int32  
 1   name             string 
 2   brand_owner      string 
 3   brand_name       string 
 4   kcal100g         float32
 5   protein100g      float32
 6   carbs100g        float32
 7   fat100g          float32
 8   unit_usdaUnitId  float64
 9   unit_name        object 
 10  unit_symbol      object 
 11  gramsPerUnit     float64
 12  displayName      object 
 13  kcal_calc        float32
 14  kcal_diff        float32
dtypes: float32(6), float64(2), int32(1), object(3), string(3)
memory usage: 148.9+ MB


In [35]:
OUTPUT_PATH = r"C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\dyeti_branded_full.csv"

df_products.to_csv(OUTPUT_PATH, index=False)

print("Zapisano CSV:")
print(OUTPUT_PATH)
print("Rozmiar (MB):", round(Path(OUTPUT_PATH).stat().st_size / (1024*1024), 2))


Zapisano CSV:
C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\dyeti_branded_full.csv
Rozmiar (MB): 323.91


# Obróbka do docelowego jsona

In [36]:
import json
from pathlib import Path

export_cols = [
    "usdaId",
    "displayName",
    "unit_usdaUnitId",
    "unit_name",
    "unit_symbol",
    "gramsPerUnit",
    "kcal100g",
    "protein100g",
    "carbs100g",
    "fat100g",
]

export_df = df_products[export_cols].copy()

def row_to_record(row):
    usda_unit_id = row["unit_usdaUnitId"]
    if pd.isna(usda_unit_id):
        usda_unit_id = None
    else:
        usda_unit_id = int(usda_unit_id)

    return {
        "usdaId": int(row["usdaId"]),
        "name": str(row["displayName"]),  # albo row["name"], jeśli chcesz bez marki
        "unit": {
            "usdaUnitId": usda_unit_id,
            "name": str(row["unit_name"]),
            "symbol": str(row["unit_symbol"]),
        },
        "gramsPerUnit": float(row["gramsPerUnit"]),
        "kcal100g": float(row["kcal100g"]),
        "protein100g": float(row["protein100g"]),
        "carbs100g": float(row["carbs100g"]),
        "fat100g": float(row["fat100g"]),
    }

records = [row_to_record(row) for _, row in export_df.iterrows()]
print(f"Liczba rekordów do eksportu: {len(records)}")

OUTPUT_JSON = Path(r"C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\dyeti_branded_full.json")

with OUTPUT_JSON.open("w", encoding="utf-8") as f:
    json.dump(records, f, ensure_ascii=False, indent=2)

print("Zapisano JSON do pliku:")
print(OUTPUT_JSON)


Liczba rekordów do eksportu: 1696793
Zapisano JSON do pliku:
C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\dyeti_branded_full.json


#### 3 Przykładowe rekordy

```json
  {
    "usdaId": 554495,
    "name": "HAMBURGERS - White Castle System Inc.",
    "unit": {
      "usdaUnitId": 1039,
      "name": "package",
      "symbol": "package"
    },
    "gramsPerUnit": 90.0,
    "kcal100g": 300.0,
    "protein100g": 13.329999923706055,
    "carbs100g": 30.0,
    "fat100g": 14.4399995803833
  },
  {
    "usdaId": 506065,
    "name": "ANTIPASTO DELIGHT - George DeLallo Co., Inc.",
    "unit": {
      "usdaUnitId": 1039,
      "name": "package",
      "symbol": "package"
    },
    "gramsPerUnit": 113.0,
    "kcal100g": 53.0,
    "protein100g": 0.0,
    "carbs100g": 7.079999923706055,
    "fat100g": 3.0999999046325684
  },
  {
    "usdaId": 506066,
    "name": "UNCURED MAPLE SMOKED HAM - WATERHILL NATURALS",
    "unit": {
      "usdaUnitId": 1050,
      "name": "slice",
      "symbol": "slice"
    },
    "gramsPerUnit": 28.0,
    "kcal100g": 107.0,
    "protein100g": 16.06999969482422,
    "carbs100g": 7.139999866485596,
    "fat100g": 3.569999933242798
  },
```

# Tworzenie mniejszego subsetu branded foods

Na cele testowe, stworzymy mniejszy zetsaw *branded foods* ( około 500 produktów ) aby nie zadusić fronta ani backendu + nie mamy budżetu NASA na deploy.

Dobór produktów do subsetu opiera się na wyborze top 30 najpopularniejszych "marek" żywienowych w Polsce ( *Coca-Cola*, *Pepsi*, *Monster Energy* itp ).

In [9]:
import pandas as pd
from pathlib import Path

# pełny dump df_products z zapisanego csv
FULL_CSV = r"C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\dyeti_branded_full.csv"

print("Wczytuję pełny df_products z CSV...")

df_products = pd.read_csv(
    FULL_CSV,
    low_memory=False,
)

print("df_products shape:", df_products.shape)
display(df_products.head(5))
df_products.info()


Wczytuję pełny df_products z CSV...
df_products shape: (1696793, 15)


,usdaId,name,brand_owner,brand_name,kcal100g,protein100g,carbs100g,fat100g,unit_usdaUnitId,unit_name,unit_symbol,gramsPerUnit,displayName,kcal_calc,kcal_diff
0,554495,HAMBURGERS,White Castle System Inc.,NaN,300.0,13.33,30.00,14.44,1039.0,package,package,90.000000,HAMBURGERS - White Castle System Inc.,303.28,3.279999
1,506065,ANTIPASTO DELIGHT,"George DeLallo Co., Inc.",NaN,53.0,0.00,7.08,3.10,1039.0,package,package,113.000000,"ANTIPASTO DELIGHT - George DeLallo Co., Inc.",56.22,3.220001
2,506066,UNCURED MAPLE SMOKED HAM,WATERHILL NATURALS,NaN,107.0,16.07,7.14,3.57,1050.0,slice,slice,28.000000,UNCURED MAPLE SMOKED HAM - WATERHILL NATURALS,124.97,17.970001
3,506068,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA",Lunds Inc,NaN,381.0,14.29,71.43,5.95,1000.0,cup,cup,168.000000,"LUNDS & BYERLYS, ORGANIC GROUND QUINOA - Lunds...",396.43,15.429993
4,506069,"SEEDS OF CHANGE, JALFREZI CURRY SAUCE, MEDIUM-HOT",Seeds Of Change Incorporated,NaN,106.0,1.18,10.59,5.88,1000.0,cup,cup,257.575758,"SEEDS OF CHANGE, JALFREZI CURRY SAUCE, MEDIUM-...",100.00,6.000000


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1696793 entries, 0 to 1696792
Data columns (total 15 columns):
 #   Column           Dtype  
---  ------           -----  
 0   usdaId           int64  
 1   name             object 
 2   brand_owner      object 
 3   brand_name       object 
 4   kcal100g         float64
 5   protein100g      float64
 6   carbs100g        float64
 7   fat100g          float64
 8   unit_usdaUnitId  float64
 9   unit_name        object 
 10  unit_symbol      object 
 11  gramsPerUnit     float64
 12  displayName      object 
 13  kcal_calc        float64
 14  kcal_diff        float64
dtypes: float64(8), int64(1), object(6)
memory usage: 194.2+ MB


### Wykrywanie popularnych marek na podstawie brand_name / brand_owner

- Tworzona jest robocza kopia `df_products` (`tmp`) oraz pomocnicze kolumny `brand_name_l` i `brand_owner_l` z nazwami marek i właścicieli w formie znormalizowanej (lowercase, string).
- Słownik `POPULAR_BRAND_PATTERNS` definiuje wzorce tekstowe dla wybranych popularnych marek (np. `monster`, `coca_cola`, `kinder`, `lays`).
- Funkcja `detect_popular_brand` łączy znormalizowane `brand_name_l` i `brand_owner_l`, a następnie sprawdza, czy któryś z wzorców marek występuje w tym tekście; zwraca odpowiedni tag (np. `"monster"`) lub `None`.
- Dla każdego produktu wyznaczany jest tag `popularBrandTag`, a następnie zliczana jest liczba produktów przypadających na poszczególne marki oraz łączna liczba produktów popularnych marek.
- Na końcu wyświetlana jest próbka (do 50 wierszy) produktów, dla których `popularBrandTag` jest niepuste, wraz z podstawowymi informacjami o nazwie i marce.


In [52]:
import pandas as pd

# kopia robocza do pracy z markami
tmp = df_products.copy()

# normalizacja brand_name / brand_owner
tmp["brand_name_l"] = tmp["brand_name"].astype("string").str.lower().fillna("")
tmp["brand_owner_l"] = tmp["brand_owner"].astype("string").str.lower().fillna("")

POPULAR_BRAND_PATTERNS = {
    "nestle": ["nestle", "nestlé"],
    "kelloggs": ["kellogg", "kellogg's"],
    "heinz": ["heinz"],
    "danone": ["danone"],
    "pepsi": ["pepsi"],
    "mars": [" mars "], 
    "coca_cola": ["coca-cola"],
    "lipton": ["lipton"],
    "oreo": ["oreo"],
    "lays": ["lay's", "lays"],
    "monster": ["monster energy", "monster beverage"],
    "pringles": ["pringles"],
    "doritos": ["doritos"],
    "cheetos": ["cheetos"],
    "kinder": ["kinder"],
    "haribo": ["haribo"],
    "red_bull": ["red bull"],
    "kitkat": ["kit kat", "kitkat"],
    "knorr": ["knorr"],
    "fanta": ["fanta"],
    "milka": ["milka"],
    "hellmanns": ["hellmann", "hellmann's"],
    "dr_oetker": ["dr. oetker", "dr oetker"],
    "snickers": ["snickers"],
    "twix": ["twix"],
    "nutella": ["nutella"],
    "nescafe": ["nescafe", "nescafé"],
    "activia": ["activia"],
    "sprite": ["sprite"],
    "nesquik": ["nesquik"],
}

def detect_popular_brand(row):
    text = (row["brand_name_l"] + " " + row["brand_owner_l"]).strip()
    if not text:
        return None
    for tag, patterns in POPULAR_BRAND_PATTERNS.items():
        for pat in patterns:
            if pat in text:
                return tag
    return None

tmp["popularBrandTag"] = tmp.apply(detect_popular_brand, axis=1)

brand_counts = tmp["popularBrandTag"].value_counts().sort_values(ascending=False)

print("Liczba produktów dla poszczególnych marek:")
print(brand_counts)

print("\nŁącznie produktów popularnych marek:",
      tmp["popularBrandTag"].notna().sum())

display(
    tmp.loc[tmp["popularBrandTag"].notna(),
            ["usdaId", "name", "brand_name", "brand_owner", "popularBrandTag"]]
       .head(30)
)


Liczba produktów dla poszczególnych marek:
popularBrandTag
nestle       16574
kelloggs     14139
heinz         9773
danone        6656
pepsi         6501
mars          5401
coca_cola     3286
lipton        1725
oreo          1517
lays          1081
monster        984
doritos        640
cheetos        539
kinder         464
haribo         409
red_bull       390
kitkat         376
knorr          232
fanta          226
milka          220
hellmanns      215
dr_oetker      123
snickers        92
twix            77
nutella         63
pringles        33
nescafe         18
activia         17
sprite          13
nesquik         13
Name: count, dtype: int64

Łącznie produktów popularnych marek: 71797


,usdaId,name,brand_name,brand_owner,popularBrandTag
382,506402,HANDMADE TROPICAL CANDY,NaN,Public Displays of Confection,lays
383,506401,TUTTI-FRUTTI CANDY,NaN,Public Displays of Confection,lays
384,506400,ROOT BEER CANDY,NaN,Public Displays of Confection,lays
542,505494,"ARROWHEAD, SPARKLING MOUNTAIN SPRING WATER, WA...",NaN,Nestle Waters North America Inc./arrowhead,nestle
564,505493,"ARROWHEAD, MOUNTAIN SPRING WATER, SPARKLING LEMON",NaN,Nestle Waters North America Inc./arrowhead,nestle
573,505490,COFFEE CREAMER,NaN,NESTLE,nestle
595,505489,COFFEE CREAMER,NaN,NESTLE,nestle
605,505568,AERO,NaN,Nestle Canada,nestle
714,505316,HOLIDAY MORSELS,NaN,Nestle USA Inc.,nestle
715,505314,MILK CHOCOLATE PEANUT BUTTER FILLED MORSEL BLO...,NaN,Nestle USA Inc.,nestle


### Selekcja produktów najpopularniejszych marek

- Definiowana jest lista `TOP_BRANDS` zawierająca wybrane popularne marki (zgodne ze znormalizowanymi tagami z `popularBrandTag`).
- Z `tmp` wybierane są tylko te produkty, które mają `popularBrandTag` należący do `TOP_BRANDS`.
- Dla każdej marki pobieranych jest maksymalnie 20 produktów (pierwsze wystąpienia), aby uzyskać reprezentatywne, ale ograniczone próbki.
- Wybrane fragmenty są łączone w jeden DataFrame (`small_df`), następnie deduplikowane po `usdaId` i resetowany jest indeks.


In [53]:
import pandas as pd

TOP_BRANDS = [
    "nestle",
    "kelloggs",
    "heinz",
    "danone",
    "pepsi",
    "mars",
    "coca_cola",
    "lipton",
    "oreo",
    "lays",
    "monster",
    "pringles",
    "doritos",
    "cheetos",
    "kinder",
    "haribo",
    "red_bull",
    "kitkat",
    "knorr",
    "fanta",
    "milka",
    "hellmanns",
    "dr_oetker",
    "snickers",
    "twix",
    "nutella",
    "nescafe",
    "activia",
    "sprite",
    "nesquik",
]

popular = tmp[tmp["popularBrandTag"].isin(TOP_BRANDS)].copy()

chunks = []
for tag in TOP_BRANDS:
    sub = popular[popular["popularBrandTag"] == tag].head(20)
    chunks.append(sub)

small_df = (
    pd.concat(chunks, ignore_index=True)
      .drop_duplicates(subset=["usdaId"])
      .reset_index(drop=True)
)

print("small_df shape (po wyborze marek):", small_df.shape)
display(small_df[["usdaId", "name", "brand_name", "brand_owner", "popularBrandTag"]].head(30))


small_df shape (po wyborze marek): (581, 18)


,usdaId,name,brand_name,brand_owner,popularBrandTag
0,505494,"ARROWHEAD, SPARKLING MOUNTAIN SPRING WATER, WA...",NaN,Nestle Waters North America Inc./arrowhead,nestle
1,505493,"ARROWHEAD, MOUNTAIN SPRING WATER, SPARKLING LEMON",NaN,Nestle Waters North America Inc./arrowhead,nestle
2,505490,COFFEE CREAMER,NaN,NESTLE,nestle
3,505489,COFFEE CREAMER,NaN,NESTLE,nestle
4,505568,AERO,NaN,Nestle Canada,nestle
5,505316,HOLIDAY MORSELS,NaN,Nestle USA Inc.,nestle
6,505314,MILK CHOCOLATE PEANUT BUTTER FILLED MORSEL BLO...,NaN,Nestle USA Inc.,nestle
7,505313,BLONDIE MIX,NaN,Nestle USA Inc.,nestle
8,505312,BITTERSWEET CHOCOLATE MORSELS,NaN,Nestle USA Inc.,nestle
9,505311,FILLED BAKING MORSELS,NaN,Nestle USA Inc.,nestle


### Finalna obróbka do json'a

Dodatkowo przeglądając mniejszą liczbe rekordów, można było zauważyć duplikaty, np *Keebler Cones Waffle Cone 216ct*. 
Duplikaty zostaną usuniete, dodatkowo finalna nazwa produktu to będZie `name` + `popular brand tag`, ponieważ będzie to krótsze oraz dość jednoznacze, do jakiej marki należy produkt. Finalnie jeszcze nazwa będzie w *lowercase*, ponieważ wszytskie nazwy produktów w *branded foods* USDA postanowiło dać samymi dużymi literami, a to wygląda nieładnie oraz nie jest spójne z obrobionymi produktami z *foundational foods*.  

Poniżek: 
- Dla każdego rekordu tworzona jest nazwa `jsonName` w formacie  
  `"nazwa_produktu - brandtag"` (całość w lowercase).  
  Jeśli produkt nie ma `popularBrandTag`, używana jest sama nazwa.
- Dane są sortowane po `usdaId`, a następnie usuwane są duplikaty po `jsonName`, tak aby każdy produkt miał unikalną nazwę.
- Sprawdzana jest obecność wymaganych kolumn (m.in. makra i informacje o jednostce). Brak którejkolwiek powoduje błąd.
- Każdy wiersz jest konwertowany do finalnej struktury JSON:  
  `usdaId`, `name`, sekcja `unit`, masa jednostki i makroskładniki na 100 g.
- Wszystkie rekordy są zapisywane do pliku `branded_foods_small.json`.


In [55]:
import json
from pathlib import Path

# budujemy nazwę: "nazwa_produktu - brandtag", wszystko lowercase
def build_json_name(row):
    base = str(row["name"]).lower().strip()
    tag = str(row.get("popularBrandTag", "") or "").lower().strip()
    tag_pretty = tag.replace("_", " ") if tag else ""
    if tag_pretty:
        return f"{base} - {tag_pretty}"
    return base

small_df["jsonName"] = small_df.apply(build_json_name, axis=1)

# usuwamy duplikaty po jsonName, zostawiamy rekord z najmniejszym usdaId
small_df = (
    small_df.sort_values("usdaId")
            .drop_duplicates(subset=["jsonName"])
            .reset_index(drop=True)
)

print("Po deduplikacji po jsonName:", small_df.shape)

# walidacja wymaganych kolumn
needed_cols = [
    "usdaId",
    "jsonName",
    "unit_usdaUnitId",
    "unit_name",
    "unit_symbol",
    "gramsPerUnit",
    "kcal100g",
    "protein100g",
    "carbs100g",
    "fat100g",
]

missing = set(needed_cols) - set(small_df.columns)
if missing:
    raise ValueError(f"Brakuje kolumn w small_df: {missing}")

def row_to_record_small(row):
    usda_unit_id = row["unit_usdaUnitId"]
    if pd.isna(usda_unit_id):
        usda_unit_id = None
    else:
        usda_unit_id = int(usda_unit_id)

    return {
        "usdaId": int(row["usdaId"]),
        "name": row["jsonName"],  
        "unit": {
            "usdaUnitId": usda_unit_id,
            "name": str(row["unit_name"]),
            "symbol": str(row["unit_symbol"]),
        },
        "gramsPerUnit": float(row["gramsPerUnit"]),
        "kcal100g": float(row["kcal100g"]),
        "protein100g": float(row["protein100g"]),
        "carbs100g": float(row["carbs100g"]),
        "fat100g": float(row["fat100g"]),
    }

records_small = [row_to_record_small(row) for _, row in small_df.iterrows()]
print("Liczba rekordów w branded_foods_small:", len(records_small))

OUTPUT_SMALL = Path(r"C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\branded_foods_small.json")
with OUTPUT_SMALL.open("w", encoding="utf-8") as f:
    json.dump(records_small, f, ensure_ascii=False, indent=2)

print("Zapisano:", OUTPUT_SMALL)


Po deduplikacji po jsonName: (471, 19)
Liczba rekordów w branded_foods_small: 471
Zapisano: C:\Users\piotr\dieti_data\dYeti_data\jowisz_2\branded_foods_small.json


## Finalne 2 przykładowe rekordy:

```json
  {
    "usdaId": 381830,
    "name": "hazelnut spread + breadsticks - nutella",
    "unit": {
      "usdaUnitId": 1059,
      "name": "unit",
      "symbol": "unit"
    },
    "gramsPerUnit": 52.0,
    "kcal100g": 519.0,
    "protein100g": 7.69,
    "carbs100g": 65.38,
    "fat100g": 26.92
  },
  {
    "usdaId": 476736,
    "name": "happy cola gummy candy - haribo",
    "unit": {
      "usdaUnitId": 1043,
      "name": "piece",
      "symbol": "piece"
    },
    "gramsPerUnit": 4.0,
    "kcal100g": 325.0,
    "protein100g": 7.5,
    "carbs100g": 72.5,
    "fat100g": 0.0
  }
```